src/stdio_mgr/types.py

```diff
@@ -0,0 +1,179 @@
+r"""``stdio_mgr.types`` *code module*.
+
+``stdio_mgr.types`` provides context managers for convenient
+interaction with ``stdin``/``stdout``/``stderr`` as a tuple.
+
+**Author**
+    John Vandenberg (jayvdb@gmail.com)
+
+**File Created**
+    6 Sep 2019
+
+**Copyright**
+    \(c) Brian Skinn 2018-2019
+
+**Source Repository**
+    http://www.github.com/bskinn/stdio-mgr
+
+**Documentation**
+    See README.rst at the GitHub repository
+
+**License**
+    The MIT License; see |license_txt|_ for full license terms
+
+**Members**
+
+"""
+import abc
+import sys
+from contextlib import ExitStack, suppress
+from io import TextIOBase
+
+import _collections_abc
+
+# AbstractContextManager was introduced in Python 3.6
+# and may be used with typing.ContextManager.
+try:
+    from contextlib import AbstractContextManager
+except ImportError:  # pragma: no cover
+
+    class AbstractContextManager(abc.ABC):
+        """An abstract base class for context managers."""
+
+        def __enter__(self):
+            """Return `self` upon entering the runtime context."""
+            return self
+
+        @abc.abstractmethod
+        def __exit__(self, exc_type, exc_value, traceback):
+            """Raise any exception triggered within the runtime context."""
+            return None
+
+        @classmethod
+        def __subclasshook__(cls, subclass):
+            """Check whether subclass is considered a subclass of this ABC."""
+            if cls is AbstractContextManager:
+                return _collections_abc._check_methods(
+                    subclass, "__enter__", "__exit__"
+                )
+            return NotImplemented
+
+
+class TupleContextManager(tuple, AbstractContextManager):
+    """Base for context managers that are also a tuple."""
+
+    # This is needed to establish a workable MRO.
+
+
+class StdioTuple(TupleContextManager):
+    """Tuple context manager of stdin, stdout and stderr stream-like objects."""
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three streams."""
+        items = list(iterable)
+        assert len(items) == 3, "iterable must be three items"  # noqa: S101
+
+        return super(StdioTuple, cls).__new__(cls, items)
+
+    @property
+    def stdin(self):
+        """Return capturing stdin stream."""
+        return self[0]
+
+    @property
+    def stdout(self):
+        """Return capturing stdout stream."""
+        return self[1]
+
+    @property
+    def stderr(self):
+        """Return capturing stderr stream."""
+        return self[2]
+
+    def _map_method(self, method: str):
```

Comment:
`_map_method` ... `suppress_all` are optional, intended to look at ways to help operate on the streams together.